C#: DataSet serialization

raulgarciamsft opened the pull request:
Queries to help potential DataSet and DataTable serialization issues.

Based on https://go.microsoft.com/fwlink/?linkid=2132227

calumgrant commented:
Many thanks for your contribution. All new queries need to be placed in the `experimental` subfolder, as detailed in https://github.com/github/codeql/blob/master/CONTRIBUTING.md#submitting-a-new-experimental-query. Would you be able to move the files (including tests) to that new location?

JarLob commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,110 @@
+      exists( Class c |
+        c.getAMethod() = this |
+        c.getABaseType*() instanceof DataSetOrTableRelatedClass or
+        c.getABaseType*() instanceof DataSetOrTableRelatedClass
```
`DataSetOrTableRelatedClass` is repeated twice

JarLob commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,110 @@
+}
+
+/**
+ * Method that may be unsafe when used to serialize DataSet and DataTable related types
```
Shouldn't it be `when used to deserialize`?

intrigus-lgtm commented:
Nice to see this query.
I don't know how important proper QLDoc is for experimental queries but it would be great if you could have a look at the [QLDoc style guide](https://github.com/github/codeql/blob/master/docs/qldoc-style-guide.md) in particular these two parts:
[QLDoc for predicates](https://github.com/github/codeql/blob/master/docs/qldoc-style-guide.md#qldoc-for-predicates)
[QLDoc for classes](https://github.com/github/codeql/blob/master/docs/qldoc-style-guide.md#qldoc-for-classes)

I don't know the policies for `qhelp` files but as they will be displayed on lgtm.com, it would probably better to integrate the relevant parts directly into the query help instead of only linking to the msft site.

raulgarciamsft commented:
> Nice to see this query.
> I don't know how important proper QLDoc is for experimental queries but it would be great if you could have a look at the [QLDoc style guide](https://github.com/github/codeql/blob/master/docs/qldoc-style-guide.md) in particular these two parts:
> [QLDoc for predicates](https://github.com/github/codeql/blob/master/docs/qldoc-style-guide.md#qldoc-for-predicates)
> [QLDoc for classes](https://github.com/github/codeql/blob/master/docs/qldoc-style-guide.md#qldoc-for-classes)
> 
> I don't know the policies for `qhelp` files but as they will be displayed on lgtm.com, it would probably better to integrate the relevant parts directly into the query help instead of only linking to the msft site.

Thanks a lot. I will fix the documentation for the classes & predicates, hopefully before the end of the day. 

Regarding the qhelp. I am actually trying to avoid duplicating any information from the forwardLink on purpose. If you think it is important that I copy some of the information, I will, but I will try to minimize duplicated data as much as I can.

raulgarciamsft commented:
Please let me know if there are any further changes needed. Thanks

JarLob commented on csharp/ql/test/experimental/Security Features/Serialization/test0.cs:
```diff
@@ -0,0 +1,101 @@
+            {
+                DataTable newTable = new DataTable();
+                System.Xml.XmlTextReader reader = new System.Xml.XmlTextReader(fs);
+                newTable.ReadXmlSchema(reader); //bug
```
AFAIK there is no bug. MS itself has a unit test that doesn't assert this case.

raulgarciamsft replied:
NOTE: Using the same .cs for multiple tests. Adding more information after `bug` to identify the test.

JarLob replied:
OK, but I would rather split the query into two. See my other comment.

JarLob commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+    this.getQualifiedName().toString() = "System.Data.DataTable.ReadXml" or
+    this.getQualifiedName().toString() = "System.Data.DataTable.ReadXmlSchema" or
+    this.getQualifiedName().toString() = "System.Data.DataSet.ReadXml" or
+    this.getQualifiedName().toString() = "System.Data.DataSet.ReadXmlSchema" or
```
IMHO `ReadXmlSchema` sink will cause false positives.
A call to `ds.ReadXml(reader, XmlReadMode.IgnoreSchema)` made after `ds.ReadXmlSchema(reader);` makes it vulnerable.
A call to `ds.ReadXml(reader)` is always vulnerable.
A call to `ds.ReadXml(reader, XmlReadMode.IgnoreSchema)` with ds that has hardcoded schema is always safe.

Maybe splitting it to queries with different certainties would be an option?

raulgarciamsft replied:
Only `cs/dataset-serialization/unsafe-type-used-data-contract-serializer` has been marked as precision = high. I can change it to medium. Would that be acceptable? Thanks

JarLob replied:
I would rather split the query into two parts. The `ReadXmlSchema` query could be low precision and `ReadXml` - medium. @hvitved  what do you think?

JarLob commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+}
+
+/**
+ * MethodCall that may be unsafe when used to serialize DataSet and DataTable related types
```
```suggestion
 * MethodCall that may be unsafe when used to deserialize DataSet and DataTable related types
```

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/DefiningPotentiallyUnsafeXmlSerializer.ql:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @name Defining a potentially unsafe XML serializer 
+ * @description Defining an XML serializable class that includes members that derive from dataSet or DataTable type may lead to a security problem. Please visit https://go.microsoft.com/fwlink/?linkid=2132227 for details.
```
dataSet -> DataSet

hvitved replied:
Not yet addressed.

raulgarciamsft replied:
fixing it

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/UnsafeTypeUsedDataContractSerializer.ql:
```diff
@@ -0,0 +1,55 @@
+  c instanceof DataSetOrTableRelatedClass
+}
+
+predicate xmlSerializerConstructorTypeParameter (Expr e) {
```
Rename to `xmlSerializerConstructorArgument`?

raulgarciamsft replied:
fixing it

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/UnsafeTypeUsedDataContractSerializer.ql:
```diff
@@ -0,0 +1,55 @@
+import csharp
+import DataSetSerialization
+
+predicate isClassDependingOnDataSetOrTable( Class c ) {
```
I would just inline the body and remove this predicate.

raulgarciamsft replied:
fixing it

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/UnsafeTypeUsedDataContractSerializer.ql:
```diff
@@ -0,0 +1,55 @@
+    e = oc.getArgument(0) |
+    c = oc.getTarget() and
+    ( 
+      c.getDeclaringType().hasQualifiedName("System.Xml.Serialization.XmlSerializer") or
```
This line is not needed.

raulgarciamsft replied:
Fixing it

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/XmlDeserializationWithDataSet.ql:
```diff
@@ -0,0 +1,16 @@
+import DataSetSerialization
+
+from UnsafeXmlReadMethodCall mc, Method m
+where m.getACall() = mc
```
`m` is not used, so may as well be removed.

raulgarciamsft replied:
Fixing it

hvitved commented on csharp/ql/test/experimental/Security Features/Serialization/options:
```diff
@@ -0,0 +1 @@
+optimize: true
```
This file can be removed.

raulgarciamsft replied:
fixed

tamasvajk commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qhelp:
```diff
@@ -0,0 +1,23 @@
+</overview>
+<recommendation>
+
+<p>Please review the <a href="https://go.microsoft.com/fwlink/?linkid=2132227">DataSet and DataTable security guidance</a> before makign use of these types for serialization.</p>
```
There's a type in this line: making vs makign.

raulgarciamsft replied:
Fixing it

raulgarciamsft commented:
> Thanks a lot for you contributions. I have a few comments. Also, could I ask you to make sure that the QL code is auto-formatted (if not, our CI tooling will complain)?

It seems like my Visual Studio deployment is missing something to automatically format ql/qll files. Any pointer to the missing component would be appreciated. Thanks

raulgarciamsft commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+  DataSetOrTable() {
+    this.getABaseType*().getQualifiedName().matches("System.Data.DataTable") or
+    this.getABaseType*().getQualifiedName().matches("System.Data.DataSet") or
+    this.getQualifiedName().matches("System.Data.DataTable") or
+    this.getQualifiedName().matches("System.Data.DataSet")
```
Fixing it.

raulgarciamsft commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+ */
+class ClassWithDataSetOrTableMember extends DataSetOrTableRelatedClass {
+  ClassWithDataSetOrTableMember() {
+    exists( Property p |
+      p = this.getAProperty() |
+      p.getType() instanceof DataSetOrTable
+    ) or exists ( AssignableMember am |
+      am = this.getAField() or
+      am = this.getAMember() |
+      am.getType() instanceof DataSetOrTable
```
Fixing it

raulgarciamsft commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+ */
+class DataSetOrTable extends DataSetOrTableRelatedClass {
+  DataSetOrTable() {
+    this.getABaseType*().getQualifiedName().matches("System.Data.DataTable") or
+    this.getABaseType*().getQualifiedName().matches("System.Data.DataSet") or
```
fixing it

raulgarciamsft commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+class SerializableClass extends Class {
+  SerializableClass() {
+    (
+      this.getABaseType*().getQualifiedName().matches("System.Xml.Serialization.XmlSerializer") or
+      this.getABaseInterface*().getQualifiedName().matches("System.Runtime.Serialization.ISerializable") or
+      this.getABaseType*().getQualifiedName().matches("System.Runtime.Serialization.XmlObjectSerializer") or
+      this.getABaseInterface*().getQualifiedName().matches("System.Runtime.Serialization.ISerializationSurrogateProvider") or
+      this.getABaseType*().getQualifiedName().matches("System.Runtime.Serialization.XmlSerializableServices") or
+      this.getABaseInterface*().getQualifiedName().matches("System.Xml.Serialization.IXmlSerializable")
```
fixing it

raulgarciamsft commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+      am = this.getAField() or
+      am = this.getAMember() |
+      am.getType() instanceof DataSetOrTable
+    ) or exists( Property p |
+      p = this.getAProperty() |
+      p.getType() instanceof DataSetOrTable or
+      p.getType().(ConstructedGeneric).getATypeArgument() instanceof DataSetOrTable 
+    )
```
I didn't though about getting anything other than collections, but given that the purpose is to flag the serialization of `DataSet`/`DataTable`, I think it is desirable to include things that are no collections as well.

raulgarciamsft commented:
> Thanks a lot for you contributions. I have a few comments. Also, could I ask you to make sure that the QL code is auto-formatted (if not, our CI tooling will complain)?

Fixed

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -29,17 +26,13 @@ class DataSetOrTable extends DataSetOrTableRelatedClass {
-      am.getType() instanceof DataSetOrTable
-    ) or exists( Property p |
-      p = this.getAProperty() |
+    exists(Property p | p = this.getAProperty() | p.getType() instanceof DataSetOrTable)
```
This line can be removed as it is included by the disjunct below (`getAMember()` includes properties).

raulgarciamsft replied:
Fixing it

raulgarciamsft commented:
I have no access to the internal details for the Language test & QHelp-preview failures, please let me know the details to make the proper changes.

raulgarciamsft commented:
The root cause for the cs/test failure was that I forgot to re-create the .expected files. Fixed.

raulgarciamsft commented:
> Thanks a lot for you contributions. I have a few comments. Also, could I ask you to make sure that the QL code is auto-formatted (if not, our CI tooling will complain)?

I figured out how to auto-format on VS Code, and I fixed the format as requested. Thanks.

I think I have fixed all or concerns. Please let me know if there is any other change I should made

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+      this.getABaseInterface*().getQualifiedName().matches("System.Xml.Serialization.IXmlSerializable")
+    ) or exists( Attribute a |
+      a = this.getAnAttribute() |
+      a.getType().getQualifiedName().toString() = "System.SerializableAttribute"
```
Not yet addressed.

raulgarciamsft replied:
fixing it

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+  SerializableClass() {
+    (
+      this.getABaseType*().getQualifiedName().matches("System.Xml.Serialization.XmlSerializer") or
+      this.getABaseInterface*().getQualifiedName().matches("System.Runtime.Serialization.ISerializable") or
```
Not yet addressed.

raulgarciamsft replied:
Fixing it

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,107 @@
+    (am = c.getAField() or am = c.getAMember()) and
+    am.getType() instanceof DataSetOrTableRelatedClass
+  )
+}
```
This can be simplified to
```
predicate isClassUnsafeXmlSerializerImplementation(SerializableClass c, AssignableMember am) {
  am = c.getAMember() and
  am.getType() instanceof DataSetOrTableRelatedClass
}
```

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,118 @@
+    this.getQualifiedName().toString() = "System.Data.DataTable.ReadXml" or
+    this.getQualifiedName().toString() = "System.Data.DataTable.ReadXmlSchema" or
+    this.getQualifiedName().toString() = "System.Data.DataSet.ReadXml" or
+    this.getQualifiedName().toString() = "System.Data.DataSet.ReadXmlSchema" or
```
Not yet addressed.

raulgarciamsft replied:
Fixing it

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/XmlDeserializationWithDataSet.ql:
```diff
@@ -0,0 +1,17 @@
+import DataSetSerialization
+
+from UnsafeXmlReadMethodCall mc
+where exists(Method m | m.getACall() = mc)
```
This line is not needed; each method call will always have a target.

raulgarciamsft replied:
I understand now. I though that you only wanted me to get rid of it on the `from` clause. Fixing it.

hvitved commented on csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll:
```diff
@@ -0,0 +1,107 @@
+/**
+ * Holds if the serializable class `c` has a property or field `m` that is of `DataSet` or `DataTable` related type
+ */
+predicate isClassUnsafeXmlSerializerImplementation(SerializableClass c, Member m) {
+  exists(Property p | m = p |
+    p = c.getAProperty() and
+    p.getType() instanceof DataSetOrTableRelatedClass
+  )
+  or
+  exists(AssignableMember am | am = m |
+    (am = c.getAField() or am = c.getAMember()) and
+    am.getType() instanceof DataSetOrTableRelatedClass
+  )
+}
```
This can be simplified to
```
predicate isClassUnsafeXmlSerializerImplementation(SerializableClass c, AssignableMember am) {
  am = c.getAMember() and
  am.getType() instanceof DataSetOrTableRelatedClass
}
```

raulgarciamsft replied:
Fixing it

raulgarciamsft replied:
Fixed both of them. But I still have no idea why the .qhelp file check is failing as I have no access to the details. Any advise will be appreciated.

hvitved replied:
The qhelp check succeeds now. There are still 4 unaddressed comments:
https://github.com/github/codeql/pull/3951#discussion_r463461083
https://github.com/github/codeql/pull/3951#discussion_r463461135
https://github.com/github/codeql/pull/3951#discussion_r463463004
https://github.com/github/codeql/pull/3951#discussion_r463463218

raulgarciamsft replied:
Fixed. Thanks, I missed those comments before.
